### PR TITLE
Deterministic Proposer Root Boosting in Forkchoice

### PR DESCRIPTION
**Pull request by rauljordan**
Currently, our proposer boost implementation relies on system time computations for its core logic. This leads to difficulty and non-determistic results when running tests. Instead, we move that responsibility to the caller of the function and significantly simplify the logic:

```go
func (f *ForkChoice) BoostProposerRoot(_ context.Context, args *forkchoicetypes.ProposerBoostRootArgs) error {
	secondsPerSlot := params.BeaconConfig().SecondsPerSlot
	isBeforeAttestingInterval := args.SecondsIntoSlot < secondsPerSlot/params.BeaconConfig().IntervalsPerSlot

	// Only update the boosted proposer root to the incoming block root
	// if the block is for the current, clock-based slot and the block was timely.
	if args.CurrentSlot == args.BlockSlot && isBeforeAttestingInterval {
		f.store.proposerBoostLock.Lock()
		f.store.proposerBoostRoot = args.BlockRoot
		f.store.proposerBoostLock.Unlock()
	}
	return nil
}
```

### CODE REVIEW

File: beacon-chain/forkchoice/doubly-linked-tree/proposer_boost_test.go [hunk 1]
```diff
@@ -429,14 +454,18 @@ func TestForkChoice_BoostProposerRoot(t *testing.T) {
 			store: &Store{},
 		}
 		// Genesis set to 1 slot ago + X where X > attesting interval.
-		genesis := time.Now().Add(-time.Duration(cfg.SecondsPerSlot) * time.Second)
-		attestingInterval := time.Duration(cfg.SecondsPerSlot / cfg.IntervalsPerSlot)
-		greaterThanAttestingInterval := attestingInterval + 100*time.Millisecond
-		genesis = genesis.Add(-greaterThanAttestingInterval * time.Second)
-		blockRoot := [32]byte{'A'}
+		attestingInterval := time.Duration(cfg.SecondsPerSlot/cfg.IntervalsPerSlot) * time.Second
+		greaterThanAttestingInterval := attestingInterval + time.Second
```
Comment (potuz) [inline beacon-chain/forkchoice/doubly-linked-tree/proposer_boost_test.go:458]: why the jump from 100 milliseconds to 1000 milliseconds?

File: beacon-chain/forkchoice/protoarray/proposer_boost_test.go [hunk 2]
```diff
@@ -426,14 +451,18 @@ func TestForkChoice_BoostProposerRoot(t *testing.T) {
 			store: &Store{},
 		}
 		// Genesis set to 1 slot ago + X where X > attesting interval.
-		genesis := time.Now().Add(-time.Duration(cfg.SecondsPerSlot) * time.Second)
-		attestingInterval := time.Duration(cfg.SecondsPerSlot / cfg.IntervalsPerSlot)
-		greaterThanAttestingInterval := attestingInterval + 100*time.Millisecond
-		genesis = genesis.Add(-greaterThanAttestingInterval * time.Second)
-		blockRoot := [32]byte{'A'}
+		attestingInterval := time.Duration(cfg.SecondsPerSlot/cfg.IntervalsPerSlot) * time.Second
+		greaterThanAttestingInterval := attestingInterval + time.Second
```
Comment (potuz) [inline beacon-chain/forkchoice/protoarray/proposer_boost_test.go:455]: same here

File: beacon-chain/forkchoice/doubly-linked-tree/proposer_boost_test.go (hunk 1 above)
Comment (rauljordan) [reply beacon-chain/forkchoice/doubly-linked-tree/proposer_boost_test.go:458]: Hey @potuz, so the proposer boost function actually operates in INTERVALS_PER_SLOT units where one interval is defined as SECONDS_PER_SLOT / 3. The logic is as follows in semi-go pseudocode:
```go
secondsPerSlot := SECONDS_PER_SLOT
isBeforeAttestingInterval := args.SecondsIntoSlot < secondsPerSlot/ INTERVALS_PER_SLOT

if args.CurrentSlot == args.BlockSlot && isBeforeAttestingInterval { 
  BOOST
}
```
as a function argument, all we care about is that we are earlier than the first interval in a slot, which is defined as unit of seconds. As such, I decided to pass in an argument of `secondsIntoSlot uint64` into the function, which makes our life easier in testing and in usage. The unit test above only worked in milliseconds, so I had to amend to seconds to make it work. The logic is the same, but now we are 1 second greater than the attesting interval rather than 100ms. We are still testing the same behavior

**Review by potuz — DISMISSED** [review]

File: beacon-chain/forkchoice/protoarray/proposer_boost_test.go [hunk 3]
```diff
@@ -163,7 +168,7 @@ func TestForkChoice_BoostProposerRoot_PreventsExAnteAttack(t *testing.T) {
 		// middle instead of the normal progression of (44 -> 34 -> 24).
 		require.Equal(t, f.store.nodes[1].weight, uint64(54))
 		require.Equal(t, f.store.nodes[2].weight, uint64(44))
-		require.Equal(t, f.store.nodes[3].weight, uint64(24))
+		require.Equal(t, f.store.nodes[3].weight, uint64(34))
```
Comment (prestonvanloon) [inline beacon-chain/forkchoice/protoarray/proposer_boost_test.go:171]: The test setup description does not match the test itself. Why has this changed?

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
Thanks for this refactoring to allow more deterministic testing. I am surprised to see that test expected outputs were changed in this PR. I would expect that this was a refactoring only and not a change in logic. Therefore, the test assertions should not have changed.

Please explain test changes in PR description so that we can review any logic changes.

File: beacon-chain/forkchoice/protoarray/proposer_boost.go [hunk 4]
```diff
@@ -19,17 +17,15 @@ import (
 //  is_before_attesting_interval = time_into_slot < SECONDS_PER_SLOT // INTERVALS_PER_SLOT
 //  if get_current_slot(store) == block.slot and is_before_attesting_interval:
 //      store.proposer_boost_root = hash_tree_root(block)
-func (f *ForkChoice) BoostProposerRoot(_ context.Context, blockSlot types.Slot, blockRoot [32]byte, genesisTime time.Time) error {
+func (f *ForkChoice) BoostProposerRoot(_ context.Context, args *types.ProposerBoostRootArgs) error {
```
Comment (prestonvanloon) [inline beacon-chain/forkchoice/protoarray/proposer_boost.go:20]: Same here, please check `args != nil`. It will probably come up in fuzzing.

File: beacon-chain/forkchoice/protoarray/proposer_boost_test.go [hunk 5]
```diff
@@ -88,8 +88,8 @@ func TestForkChoice_BoostProposerRoot_PreventsExAnteAttack(t *testing.T) {
 		//         2
 		//         |
 		//         3 <- HEAD
-		slot = types.Slot(2)
-		newRoot = indexToHash(2)
+		slot = types.Slot(3)
+		newRoot = indexToHash(3)
```
Comment (prestonvanloon) [inline beacon-chain/forkchoice/protoarray/proposer_boost_test.go:92]: Why this change?

File: beacon-chain/forkchoice/protoarray/proposer_boost_test.go [hunk 6]
```diff
@@ -128,15 +128,20 @@ func TestForkChoice_BoostProposerRoot_PreventsExAnteAttack(t *testing.T) {
...
+			SecondsIntoSlot: 0,
+		}
+		require.NoError(t, f.BoostProposerRoot(ctx, args))
 		headRoot, err = f.Head(ctx, jEpoch, zeroHash, balances, fEpoch)
 		require.NoError(t, err)
 		assert.Equal(t, newRoot, headRoot, "Incorrect head for justified epoch at slot 3")
 
 		// Check the ancestor scores from the store.
-		require.Equal(t, 4, len(f.store.nodes))
+		require.Equal(t, 5, len(f.store.nodes))
```
Comment (prestonvanloon) [inline beacon-chain/forkchoice/protoarray/proposer_boost_test.go:144]: This seems like a logic change as well. Why is this different?

File: beacon-chain/forkchoice/protoarray/proposer_boost_test.go (hunk 5 above)
Comment (rauljordan) [reply beacon-chain/forkchoice/protoarray/proposer_boost_test.go:92]: This test case in particular was not matching the comments, and not caught in coverage because it doesn't make a difference to proposer boost if we did slot 2 or 3, but my comments say it should be slot 3. The behavior tested is still the same, but made it 3 to align more with my comment

File: beacon-chain/forkchoice/protoarray/proposer_boost_test.go (hunk 6 above)
Comment (rauljordan) [reply beacon-chain/forkchoice/protoarray/proposer_boost_test.go:144]: Because in line 91 we were overriding the node at slot 2 when we should have been inserting one at slot 3, so there should have actually been 5 nodes this whole time when I was asserting 4

File: beacon-chain/forkchoice/protoarray/proposer_boost_test.go (hunk 3 above)
Comment (rauljordan) [reply beacon-chain/forkchoice/protoarray/proposer_boost_test.go:176]: Hi @prestonvanloon my test assertions were actually wrong here! The reason is explained below. We were inserting a node at slot 2 when I was intending to insert at slot 3 instead

**Comment by rauljordan** [comment]
This is ready again @prestonvanloon

**Review by prestonvanloon — APPROVED** [review]
